src/shared/tests.c

```diff
@@ -41,6 +42,16 @@ const char* get_exe_relative_testdata_dir(void) {
         static char testdir[PATH_MAX];
 
         assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
+
+        /* Check if we're running from the builddir. If so, use the compiled in path. */
+        if (path_startswith(exedir, ABS_BUILD_DIR))
+                return ABS_SRC_DIR "/test";
+
+        /* Try relative path, according to the install-test layout */
         assert_se(snprintf(testdir, sizeof(testdir), "%s/testdata", dirname(exedir)) > 0);
-        return testdir;
+        if (access(testdir, F_OK) >= 0)
+                return testdir;
+
+        fputs("ERROR: Cannot find testdata directory, set $SYSTEMD_TEST_DATA\n", stderr);
```

Comment:
I mean setting S_D_T to a non-existent directory.

`S_D_T=totally-wrong-directory` will produce strange output. I think that's OK. (I have no idea how to check correctness of the S_D_T-contents :))